### DISCUSSION TITLE
Finding the bottom of the table for strokes

### DISCUSSION THREAD

**Discussion by StuieT85** [typst/typst#3692] [General]
### Description

Is it possible to find the bottom row of a table when setting the strokes for a table. 

### Use Case

I am trying to create a table that has strokes in the first row, top and bottom, and at the bottom of the table. I used the guide which had a similar example setting the top stroke if y < 2 and the bottom stroke always,  but this breaks if there is a page break as the header row on the second page is not proceeded by the first row so on the second page does not get its bottom border.

I have found that setting the header row to top and bottom stroke and then manually setting the bottom row to a bottom stroke but this relies on me knowing how many rows are in the table.

```typst
#set page(height: 10em, width: 25em )

#set table(
    columns: (1fr, 1fr, 1fr, 1fr),
  stroke: (_, y) => {
    if y == 0  {(
      top: color.black,
      bottom: color.black
    )} else if y == 6 {(
      bottom: color.black
    )}
  }
)

#table(
  table.header[Header 1][Header 2][Header 3][Header 4],
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
)
```
I have looked through the table and grid documentation and the grid documentation suggests there is a `bottom` that can be referenced to reference the bottom row but I have tried using it and, while it does not cause an error, does not result in a bottom border

```typst
#set page(height: 10em, width: 25em )

#set table(
    columns: (1fr, 1fr, 1fr, 1fr),
  stroke: (_, y) => {
    if y == 0  {(
      top: color.black,
      bottom: color.black
    )} else if y == bottom {(
      bottom: color.black
    )}
  }
)

#table(
  table.header[Header 1][Header 2][Header 3][Header 4],
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
)
```
I apologise if I am missing something obvious

**Comment by frozolotl** [comment]
You could use the following solution:
```typ
#set page(height: 10em, width: 25em )

#set table(stroke: (_, y) => if y == 0 { (bottom: 1pt) })
#show table: block.with(stroke: (y: 1pt))

#table(
  columns: (1fr,) * 4,
  table.header[Header 1][Header 2][Header 3][Header 4],
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
)
```

I don't think this solution is totally obvious or ideal, though.
CC: @PgBiel

**Comment by nmayorov** [comment]
I was dealing with the same question.

Having something like "-1" or "last" would be ideal to make it fully functional and straightforward.

Solution with "block" does seem to have some limitations, I can't apply arbitrary styling with it.

**Comment by PgBiel** [comment]
For a single table, you can add a `table.hline` at the very bottom, e.g.

```typ
#table(
  [a], [b],
  [c], [d],
  table.hline(stroke: red)
)
```

This won't work for templates, however (unless you redefine the `table` function to always add that hline at the bottom).

It is expected to have such a feature for table templates in the future.

**Comment by eriol** [comment]
I wanted to do the same and I ended with:
```typst
#set page(height: 10em, width: 25em )

#set table(
    columns: (1fr, 1fr, 1fr, 1fr),
    stroke: (_, y) => (
       top: if y <= 1 { 1pt } else { 0pt },
       bottom: 1pt,
   )
)

#table(
  table.header[Header 1][Header 2][Header 3][Header 4],
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
  "Contents A", "Contents B", "Contents C", "Contents D",
)
```

This give us:
![table1](https://github.com/typst/typst/assets/212279/ba5720a8-6108-41c6-9cb4-f5f0f7345123)

**Comment by cskeeters** [comment]
![Three-line table](https://uviclibraries.github.io/latex/images/act-2/refer-table.png)

Iif you want to go for a [research-paper style](https://uviclibraries.github.io/latex/act-2.html), you can use this:

```typst
#set table(
    // none can be overridden, but 0pt can't
    stroke: (_, y) => (
        left: { 0pt },
        right: { 0pt },
        top: if y < 1 { stroke(1pt) } else if y == 1 { none } else { 0pt },
        bottom: if y < 1 { stroke(.5pt) } else { stroke(1pt) },
    ),
)

#show figure.where(kind:table): set figure.caption(position: top)
```

**Reply by benediktschfr** [reply]
> ```typst
> #set table(
>     // none can be overridden, but 0pt can't
>     stroke: (_, y) => (
>         left: { 0pt },
>         right: { 0pt },
>         top: if y < 1 { stroke(1pt) } else if y == 1 { none } else { 0pt },
>         bottom: if y < 1 { stroke(.5pt) } else { stroke(1pt) },
>     ),
> )
> 
> #show figure.where(kind:table): set figure.caption(position: top)
> ```

And which could be simplified to
```typst
 #set table(
  stroke: (_, y) => (
    top: if y <= 1 { 1pt } else { 0pt },
    bottom: 1pt,
  )
)
```
if stroke thickness may be the same in all rows.
